[VectorTypeUtil] Support struct return types for widen intrinsics (NFC).

arcbbb opened the pull request:
Splitting from #151300,  vp_load_ff returns a struct type that cannot be widened by toVectorizedTy.
This patch adds isVectorIntrinsicWithStructReturnScalarAtField and widen each struct element type independently.

arcbbb commented:
ping

MacDue commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4100,7 +4100,22 @@ static bool willGenerateVectors(VPlan &Plan, ElementCount VF,
+          Tys.push_back(ElementTy);
+        }
+        WideTy = StructType::create(Tys);
+      } else
```
This should be part of the `toVectorizedTy` helper, as otherwise the helper is pointless as it can't return a vectorized type.

arcbbb commented on llvm/include/llvm/Analysis/VectorUtils.h:
```diff
@@ -164,6 +164,11 @@ LLVM_ABI bool
 
+/// Identifies if the vector form of the intrinsic that returns a struct has
+/// a scalar element at the struct element index \p RetIdx.
+LLVM_ABI bool isVectorIntrinsicWithStructReturnScalarAtField(Intrinsic::ID ID,
```
Intrinsic helpers live in Analysis/VectorUtils.h, so shoving this into IR/VectorTypeUtils.* would be a layering headache.
Let me rethink other options.

arcbbb commented on llvm/lib/IR/VectorTypeUtils.cpp:
```diff
@@ -8,23 +8,35 @@
+        StructTy->getContext(),
+        map_to_vector(StructTy->elements(), [&](Type *ElTy) -> Type * {
+          return VectorType::get(ElTy, EC);
+        }));
```
Updated. Thanks!

MacDue commented on llvm/lib/IR/VectorTypeUtils.cpp:
```diff
@@ -27,6 +28,30 @@ Type *llvm::toVectorizedStructTy(StructType *StructTy, ElementCount EC) {
+              isVectorIntrinsicWithStructReturnScalarAtField(IID, It.index()))
+            return ElTy;
+          return VectorType::get(ElTy, EC);
+        }));
```
I think this code can still just go in `toVectorizedStructTy`, adding an `IID` argument (with a default of 0), rather than duplicating the same logic twice. This function can then just call `toVectorizedStructTy` with an `IID`.

arcbbb commented on llvm/lib/IR/VectorTypeUtils.cpp:
```diff
@@ -27,6 +28,30 @@ Type *llvm::toVectorizedStructTy(StructType *StructTy, ElementCount EC) {
+        }));
+  }
+  if (IID != 0 && isVectorIntrinsicWithScalarOpAtArg(IID, -1, nullptr))
+    return Ty;
```
No, it’s just defensive if the return type isn’t a struct or a vector.

arcbbb commented:
gentle ping

fhahn commented on llvm/lib/IR/VectorTypeUtils.cpp:
```diff
@@ -22,7 +34,11 @@ Type *llvm::toVectorizedStructTy(StructType *StructTy, ElementCount EC) {
+      map_to_vector(enumerate(StructTy->elements()), [&](auto It) -> Type * {
+        Type *ElTy = It.value();
+        if (IID != 0 &&
+            isVectorIntrinsicWithStructReturnScalarAtField(IID, It.index()))
```
```suggestion
        if (
            isVectorIntrinsicWithStructReturnScalarAtField(IID, It.index()))
```

IID ==0 should naturally be handled by `isVectorIntrinsicWithStructReturnScalarAtField`

fhahn commented on llvm/include/llvm/IR/VectorTypeUtils.h:
```diff
@@ -14,6 +14,11 @@
 namespace llvm {
 
+/// Identifies if the vector form of the intrinsic that returns a struct has
+/// a scalar element at the struct element index \p EleIdx.
```
```suggestion
/// Returns true if \p ID is a vector intrinsic that returns a struct with a scalar element at index \p EleIdx.
```

arcbbb replied:
Updated. Thanks!

fhahn commented on llvm/include/llvm/IR/VectorTypeUtils.h:
```diff
@@ -14,6 +14,11 @@
 
+/// Identifies if the vector form of the intrinsic that returns a struct has
+/// a scalar element at the struct element index \p EleIdx.
+LLVM_ABI bool isVectorIntrinsicWithStructReturnScalarAtField(unsigned ID,
```
```suggestion
LLVM_ABI bool isVectorIntrinsicWithStructReturnScalarAtField(unsigned IID,
```
for consistency with change sblow

arcbbb committed:
[VectorTypeUtil] Support struct return types for widen intrinsics (NFC). #165218

Splitting from #151300, vp_load_ff returns a struct type that cannot be
widened by toVectorizedTy.
This patch adds isVectorIntrinsicWithStructReturnScalarAtField and widen
each struct element type independently.

arcbbb committed:
Split LV changes

arcbbb commented on llvm/lib/IR/VectorTypeUtils.cpp:
```diff
@@ -8,12 +8,24 @@
+    return EleIdx == 1;
+  default:
+    return false;
+  }
```
Updated. Thanks!

arcbbb commented on llvm/lib/Analysis/VectorUtils.cpp:
```diff
@@ -224,6 +224,10 @@ bool llvm::isVectorIntrinsicWithStructReturnOverloadAtField(
+  case Intrinsic::modf:
+  case Intrinsic::sincos:
+  case Intrinsic::sincospi:
+    return false;
```
These intrinsics are found in vectorizer tests. 
Remove for now. We'll add them when isVectorIntrinsicWithStructReturnOverloadAtField is adopted in VPWidenIntrinsicRecipe::execute.

arcbbb committed:
Merge branch 'main' into handle-ret-struct
